src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+    
+                // ** P **
+                float pidP = latLong->distance * positionPidCoeffs.Kp;
+    
+                // ** I **
+                if (!latLong->isStarting){
+                    // only accumulate iTerm after completing the start phase
+                    // perhaps need a timeout on the start phase ?
+                    latLong->integral += latLong->distance * posHold.gpsDataIntervalS;
+                } else {
+                    // while moving sticks, slowly leak iTerm away, approx 2s time constant
+                    latLong->integral *= leak;
+                }
+                float pidI = latLong->integral * positionPidCoeffs.Ki;
+        
+                // ** D ** //
+                // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+                // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
+                float velocity = (latLong->distance - latLong->previousDistance) * posHold.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
+                latLong->previousDistance = latLong->distance;
+    
+                float acceleration = (velocity - latLong->previousVelocity) * posHold.gpsDataFreqHz;
+                latLong->previousVelocity = velocity;
+        
+                // scale and filter - filter cutoffs vary during the startup phase
+                float pidD = velocity * positionPidCoeffs.Kd;
+                pt1FilterUpdateCutoff(&latLong->velocityLpf, posHold.pt1Gain);
+                pidD = pt1FilterApply(&latLong->velocityLpf, pidD);
+    
+                float pidA = acceleration * positionPidCoeffs.Kd;
+                pt2FilterUpdateCutoff(&latLong->accelerationLpf, posHold.pt1Gain);
+                pidA = pt2FilterApply(&latLong->accelerationLpf, pidA);
+        
+                // limit sum of D and A because otherwise can be too aggressive when starting at speed
+                const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+                // an angle of more than 35 degrees is achieved as P and I grow
+                // ** todo = should this be half of the user-configurable angle_limit?  Or fixed?
+                const float pidDA = constrainf(pidD + pidA, -maxDAAngle, maxDAAngle);
+    
+                // ** PID Sum **
+                float pidSum = pidP + pidI + pidDA;
+    
+                // terminate initial startup behaviour separately for latitude and longitude controllers
+                // the position target is reset when pidSum crosses zero
+                // this enhances the smoothness of the transition from stick input back to position hold - there is no sharp change in pidSum
+                if (latLong->isStarting && latLong->pidSum * pidSum < 0.0f) { // pidsum ns has reversed sign
```

Comment:
Ad 2d control) testing 2d vector for zero crossing is problematic. But the test can 'slowed to reasonable speed' or similar. 

This way, 2 points are captured partially independently - NS and EW. The is time interval when only one axis is fixed, possibly leading to problems

Reply:
I made a detailed response a bit further up.  Yes there are a number of methods we could use.  This one proved to be smoothest - more smooth than looking for a minimum abs velocity - and as reliable as any other method I tested.